docs/ci_getnameinfo.3
.\"
.\" Copyright 2005 by Dominick Meglio.
.\" SPDX-License-Identifier: MIT
.\"
.TH CI_GETNAMEINFO 3 "1 May 2009"
.SH NAME
ci_getnameinfo \- Address-to-nodename translation in protocol-independent manner
.SH SYNOPSIS
.nf
#include <ci.h>

typedef void (*ci_nameinfo_callback)(void *\fIarg\fP, int \fIstatus\fP,
                                       int \fItimeouts\fP, char *\fInode\fP,
                                       char *\fIservice\fP)

void ci_getnameinfo(ci_channel_t *\fIchannel\fP, const struct sockaddr *\fIsa\fP,
                      ci_socklen_t \fIsalen\fP, int \fIflags\fP,
                      ci_nameinfo_callback \fIcallback\fP, void *\fIarg\fP)
.fi
.SH DESCRIPTION
The
.B ci_getnameinfo
function is defined for protocol-independent address translation. The function
is a combination of \fIci_gethostbyaddr(3)\fP and \fIgetservbyport(3)\fP. The function will
translate the address either by executing a host query on the name service channel
identified by
.IR channel 
or it will attempt to resolve it locally if possible.
The parameters
.I sa
and
.I len
give the address as a sockaddr structure, and
.I flags
gives the options that the function will use.  Valid flags are listed below:
.TP 19
.B CI_NI_NOFQDN
Only the nodename portion of the FQDN is returned for local hosts.
.TP 19
.B CI_NI_NUMERICHOST
The numeric form of the hostname is returned rather than the name.
.TP 19
.B CI_NI_NAMEREQD
An error is returned if the hostname cannot be found in the DNS.
.TP 19
.B CI_NI_NUMERICSERV
The numeric form of the service is returned rather than the name.
.TP 19
.B CI_NI_TCP
The service name is to be looked up for the TCP protocol.
.TP 19
.B CI_NI_UDP
The service name is to be looked up for the UDP protocol.
.TP 19
.B CI_NI_SCTP
The service name is to be looked up for the SCTP protocol.
.TP 19
.B CI_NI_DCCP
The service name is to be looked up for the DCCP protocol.
.TP 19
.B CI_NI_NUMERICSCOPE
The numeric form of the scope ID is returned rather than the name.
.TP 19
.B CI_NI_LOOKUPHOST
A hostname lookup is being requested.
.TP 19
.B CI_NI_LOOKUPSERVICE
A service name lookup is being requested.
.PP
When the query
is complete or has 
failed, the ci library will invoke \fIcallback\fP.  Completion or failure of 
the query may happen immediately, or may happen during a later call to
\fIci_process(3)\fP, \fIci_destroy(3)\fP or \fIci_cancel(3)\fP.
.PP
When the associated callback is called, it is called with a channel lock so care
must be taken to ensure any processing is minimal to prevent DNS channel stalls.

The callback may be triggered from a different thread than the one which
called \fIci_getnameinfo(3)\fP.

For integrators running their own event loops and not using \fBCI_OPT_EVENT_THREAD\fP,
care needs to be taken to ensure any file descriptor lists are updated immediately
within the eventloop when notified.
.PP
The callback argument
.I arg
is copied from the
.B ci_getnameinfo
argument
.IR arg .
The callback argument
.I status
indicates whether the query succeeded and, if not, how it failed.  It
may have any of the following values:
.TP 19
.B CI_SUCCESS
The host lookup completed successfully.
.TP 19
.B CI_ENOTIMP
The ci library does not know how to look up addresses of type
.IR family .
.TP 19
.B CI_ENOTFOUND
The address
.I addr
was not found.
.TP 19
.B CI_ENOMEM
Memory was exhausted.
.TP 19
.B CI_ECANCELLED
The query was cancelled.
.TP 19
.B CI_EDESTRUCTION
The name service channel
.I channel
is being destroyed; the query will not be completed.
.TP 19
.B CI_EBADFLAGS
The
.I flags
parameter contains an illegal value.
.PP
The callback argument
.I timeouts
reports how many times a query timed out during the execution of the
given request.
.PP
On successful completion of the query, the callback argument
.I node
contains a string representing the hostname (assuming 
.B CI_NI_LOOKUPHOST
was specified). Additionally, 
.I service
contains a string representing the service name (assuming
.B CI_NI_LOOKUPSERVICE
was specified).
If the query did not complete successfully, or one of the values
was not requested, 
.I node
or
.I service
will be 
.BR NULL .
.SH SEE ALSO
.BR ci_process (3),


RELEASE-NOTES.md
## c-ci version 1.34.2 - October 15 2024

This release contains a fix for downstream packages detecting the c-ci
version based on the contents of the header file rather than the
distributed pkgconf or cmake files.

## c-ci version 1.34.1 - October 9 2024

This release fixes a packaging issue.


## c-ci version 1.34.0 - October 9 2024

This is a feature and bugfix release.

Features:
* adig: read arguments from adigrc.
  [PR #856](https://github.com/c-ci/c-ci/pull/856)
* Add new pending write callback optimization via `ci_set_pending_write_cb`.
  [PR #857](https://github.com/c-ci/c-ci/pull/857)
* New function `ci_process_fds()`.
  [PR #875](https://github.com/c-ci/c-ci/pull/875)
* Failed servers should be probed rather than redirecting queries which could
  cause unexpected latency.
  [PR #877](https://github.com/c-ci/c-ci/pull/877)
* adig: rework command line arguments to mimic dig from bind.
  [PR #890](https://github.com/c-ci/c-ci/pull/890)
* Add new method for overriding network functions
  `ci_set_socket_function_ex()` to properly support all new functionality.
  [PR #894](https://github.com/c-ci/c-ci/pull/894)
* Fix regression with custom socket callbacks due to DNS cookie support.
  [PR #895](https://github.com/c-ci/c-ci/pull/895)
* ci_socket: set IP_BIND_ADDRESS_NO_PORT on ci_set_local_ip* tcp sockets
  [PR #887](https://github.com/c-ci/c-ci/pull/887)
* URI parser/writer for ci_set_servers_csv()/ci_get_servers_csv().
  [PR #882](https://github.com/c-ci/c-ci/pull/882)

Changes:
* Connection handling modularization.
  [PR #857](https://github.com/c-ci/c-ci/pull/857),
  [PR #876](https://github.com/c-ci/c-ci/pull/876)
* Expose library/utility functions to tools.
  [PR #860](https://github.com/c-ci/c-ci/pull/860)
* Remove `ci__` prefix, just use `ci_` for internal functions.
  [PR #872](https://github.com/c-ci/c-ci/pull/872)


Bugfixes:
* fix: potential WIN32_LEAN_AND_MEAN redefinition.
  [PR #869](https://github.com/c-ci/c-ci/pull/869)
* Fix googletest v1.15 compatibility.
  [PR #874](https://github.com/c-ci/c-ci/pull/874)
* Fix pkgconfig thread dependencies.
  [PR #884](https://github.com/c-ci/c-ci/pull/884)


Thanks go to these friendly people for their efforts and contributions for this
release:

* Brad House (@bradh352)
* Cristian Rodríguez (@crrodriguez)
* Georg (@tacerus)
* @lifenjoiner
* Shelley Vohr (@codebytere)
* 前进，前进，进 (@leleliu008)



src/lib/ci_getnameinfo.c
/* MIT License
 *
 * Copyright (c) 2005, 2013 Dominick Meglio
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice (including the next
 * paragraph) shall be included in all copies or substantial portions of the
 * Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 *
 * SPDX-License-Identifier: MIT
 */
#include "ci_private.h"

#ifdef HAVE_GETSERVBYPORT_R
#  if !defined(GETSERVBYPORT_R_ARGS) || (GETSERVBYPORT_R_ARGS < 4) || \
    (GETSERVBYPORT_R_ARGS > 6)
#    error "you MUST specify a valid number of arguments for getservbyport_r"
#  endif
#endif

#ifdef HAVE_NETINET_IN_H
#  include <netinet/in.h>
#endif
#ifdef HAVE_NETDB_H
#  include <netdb.h>
#endif
#ifdef HAVE_ARPA_INET_H
#  include <arpa/inet.h>
#endif

#include "ci_nameser.h"

#ifdef HAVE_NET_IF_H
#  include <net/if.h>
#endif
#if defined(USE_WINSOCK) && defined(HAVE_IPHLPAPI_H)
#  include <iphlpapi.h>
#endif

#include "ci_ipv6.h"

struct nameinfo_query {
  ci_nameinfo_callback callback;
  void                  *arg;

  union {
    struct sockaddr_in  addr4;
    struct sockaddr_in6 addr6;
  } addr;

  int          family;
  unsigned int flags;
  size_t       timeouts;
};

#ifdef HAVE_STRUCT_SOCKADDR_IN6_SIN6_SCOPE_ID
#  define IPBUFSIZ \
    (sizeof("ffff:ffff:ffff:ffff:ffff:ffff:255.255.255.255") + IF_NAMESIZE)
#else
#  define IPBUFSIZ (sizeof("ffff:ffff:ffff:ffff:ffff:ffff:255.255.255.255"))
#endif

static void  nameinfo_callback(void *arg, int status, int timeouts,
                               struct hostent *host);
static char *lookup_service(unsigned short port, unsigned int flags, char *buf,
                            size_t buflen);
#ifdef HAVE_STRUCT_SOCKADDR_IN6_SIN6_SCOPE_ID
static void append_scopeid(const struct sockaddr_in6 *addr6, unsigned int flags,
                           char *buf, size_t buflen);
#endif
static char *ci_striendstr(const char *s1, const char *s2);

static void  ci_getnameinfo_int(ci_channel_t        *channel,
                                  const struct sockaddr *sa,
                                  ci_socklen_t salen, int flags_int,
                                  ci_nameinfo_callback callback, void *arg)
{
  const struct sockaddr_in  *addr  = NULL;
  const struct sockaddr_in6 *addr6 = NULL;
  struct nameinfo_query     *niquery;
  unsigned short             port  = 0;
  unsigned int               flags = (unsigned int)flags_int;

  /* Validate socket address family and length */
  if (sa && sa->sa_family == AF_INET &&
      salen >= (ci_socklen_t)sizeof(struct sockaddr_in)) {
    addr = CI_INADDR_CAST(const struct sockaddr_in *, sa);
    port = addr->sin_port;
  } else if (sa && sa->sa_family == AF_INET6 &&
             salen >= (ci_socklen_t)sizeof(struct sockaddr_in6)) {
    addr6 = CI_INADDR_CAST(const struct sockaddr_in6 *, sa);
    port  = addr6->sin6_port;
  } else {
    callback(arg, CI_ENOTIMP, 0, NULL, NULL);
    return;
  }

  /* If neither, assume they want a host */
  if (!(flags & CI_NI_LOOKUPSERVICE) && !(flags & CI_NI_LOOKUPHOST)) {
    flags |= CI_NI_LOOKUPHOST;
  }

  /* All they want is a service, no need for DNS */
  if ((flags & CI_NI_LOOKUPSERVICE) && !(flags & CI_NI_LOOKUPHOST)) {
    char  buf[33];
    char *service;

    service =
      lookup_service((unsigned short)(port & 0xffff), flags, buf, sizeof(buf));
    callback(arg, CI_SUCCESS, 0, NULL, service);
    return;
  }

  /* They want a host lookup */
  if (flags & CI_NI_LOOKUPHOST) {
    /* A numeric host can be handled without DNS */
    if (flags & CI_NI_NUMERICHOST) {
      char  ipbuf[IPBUFSIZ];
      char  srvbuf[33];
      char *service = NULL;
      ipbuf[0]      = 0;

      /* Specifying not to lookup a host, but then saying a host
       * is required has to be illegal.
       */
      if (flags & CI_NI_NAMEREQD) {
        callback(arg, CI_EBADFLAGS, 0, NULL, NULL);
        return;
      }
      if (sa->sa_family == AF_INET6) {
        ci_inet_ntop(AF_INET6, &addr6->sin6_addr, ipbuf, IPBUFSIZ);
        /* If the system supports scope IDs, use it */
#ifdef HAVE_STRUCT_SOCKADDR_IN6_SIN6_SCOPE_ID
        append_scopeid(addr6, flags, ipbuf, sizeof(ipbuf));
#endif
      } else {
        ci_inet_ntop(AF_INET, &addr->sin_addr, ipbuf, IPBUFSIZ);
      }
      /* They also want a service */
      if (flags & CI_NI_LOOKUPSERVICE) {
        service = lookup_service((unsigned short)(port & 0xffff), flags, srvbuf,
                                 sizeof(srvbuf));
      }
      callback(arg, CI_SUCCESS, 0, ipbuf, service);
      return;
    } else {
      /* This is where a DNS lookup becomes necessary */
      niquery = ci_malloc(sizeof(struct nameinfo_query));
      if (!niquery) {
        callback(arg, CI_ENOMEM, 0, NULL, NULL);
        return;
      }
      niquery->callback = callback;
      niquery->arg      = arg;
      niquery->flags    = flags;
      niquery->timeouts = 0;
      if (sa->sa_family == AF_INET) {
        niquery->family = AF_INET;
        memcpy(&niquery->addr.addr4, addr, sizeof(niquery->addr.addr4));
        ci_gethostbyaddr_nolock(channel, &addr->sin_addr,
                                  sizeof(struct in_addr), AF_INET,
                                  nameinfo_callback, niquery);
      } else {
        niquery->family = AF_INET6;
        memcpy(&niquery->addr.addr6, addr6, sizeof(niquery->addr.addr6));
        ci_gethostbyaddr_nolock(channel, &addr6->sin6_addr,
                                  sizeof(struct ci_in6_addr), AF_INET6,
                                  nameinfo_callback, niquery);
      }
    }
  }
}

void ci_getnameinfo(ci_channel_t *channel, const struct sockaddr *sa,
                      ci_socklen_t salen, int flags_int,
                      ci_nameinfo_callback callback, void *arg)
{
  if (channel == NULL) {
    return;
  }

  ci_channel_lock(channel);
  ci_getnameinfo_int(channel, sa, salen, flags_int, callback, arg);
  ci_channel_unlock(channel);
}

static void nameinfo_callback(void *arg, int status, int timeouts,
                              struct hostent *host)
{
  struct nameinfo_query *niquery = (struct nameinfo_query *)arg;
  char                   srvbuf[33];
  char                  *service = NULL;

  niquery->timeouts += (size_t)timeouts;
  if (status == CI_SUCCESS) {
    /* They want a service too */
    if (niquery->flags & CI_NI_LOOKUPSERVICE) {
      if (niquery->family == AF_INET) {
        service = lookup_service(niquery->addr.addr4.sin_port, niquery->flags,
                                 srvbuf, sizeof(srvbuf));
      } else {
        service = lookup_service(niquery->addr.addr6.sin6_port, niquery->flags,
                                 srvbuf, sizeof(srvbuf));
      }
    }
    /* NOFQDN means we have to strip off the domain name portion.  We do
       this by determining our own domain name, then searching the string
       for this domain name and removing it.
     */
#ifdef HAVE_GETHOSTNAME
    if (niquery->flags & CI_NI_NOFQDN) {
      char        buf[255];
      const char *domain;
      gethostname(buf, 255);
      if ((domain = strchr(buf, '.')) != NULL) {
        char *end = ci_striendstr(host->h_name, domain);
        if (end) {
          *end = 0;
        }
      }
    }
#endif
    niquery->callback(niquery->arg, CI_SUCCESS, (int)niquery->timeouts,
                      host->h_name, service);
    ci_free(niquery);
    return;
  }
  /* We couldn't find the host, but it's OK, we can use the IP */
  else if (status == CI_ENOTFOUND && !(niquery->flags & CI_NI_NAMEREQD)) {
    char ipbuf[IPBUFSIZ];
    if (niquery->family == AF_INET) {
      ci_inet_ntop(AF_INET, &niquery->addr.addr4.sin_addr, ipbuf, IPBUFSIZ);
    } else {
      ci_inet_ntop(AF_INET6, &niquery->addr.addr6.sin6_addr, ipbuf, IPBUFSIZ);
#ifdef HAVE_STRUCT_SOCKADDR_IN6_SIN6_SCOPE_ID
      append_scopeid(&niquery->addr.addr6, niquery->flags, ipbuf,
                     sizeof(ipbuf));
#endif
    }
    /* They want a service too */
    if (niquery->flags & CI_NI_LOOKUPSERVICE) {
      if (niquery->family == AF_INET) {
        service = lookup_service(niquery->addr.addr4.sin_port, niquery->flags,
                                 srvbuf, sizeof(srvbuf));
      } else {
        service = lookup_service(niquery->addr.addr6.sin6_port, niquery->flags,
                                 srvbuf, sizeof(srvbuf));
      }
    }
    niquery->callback(niquery->arg, CI_SUCCESS, (int)niquery->timeouts, ipbuf,
                      service);
    ci_free(niquery);
    return;
  }
  niquery->callback(niquery->arg, status, (int)niquery->timeouts, NULL, NULL);
  ci_free(niquery);
}

static char *lookup_service(unsigned short port, unsigned int flags, char *buf,
                            size_t buflen)
{
  const char     *proto;
  struct servent *sep;
#ifdef HAVE_GETSERVBYPORT_R
  struct servent se;
#endif
  char        tmpbuf[4096];
  const char *name;
  size_t      name_len;

  if (port) {
    if (flags & CI_NI_NUMERICSERV) {
      sep = NULL;
    } else {
      if (flags & CI_NI_UDP) {
        proto = "udp";
      } else if (flags & CI_NI_SCTP) {
        proto = "sctp";
      } else if (flags & CI_NI_DCCP) {
        proto = "dccp";
      } else {
        proto = "tcp";
      }
#ifdef HAVE_GETSERVBYPORT_R
      memset(&se, 0, sizeof(se));
      sep = &se;
      memset(tmpbuf, 0, sizeof(tmpbuf));
#  if GETSERVBYPORT_R_ARGS == 6
      if (getservbyport_r(port, proto, &se, (void *)tmpbuf, sizeof(tmpbuf),
                          &sep) != 0) {
        sep = NULL; /* LCOV_EXCL_LINE: buffer large so this never fails */
      }
#  elif GETSERVBYPORT_R_ARGS == 5
      sep = getservbyport_r(port, proto, &se, (void *)tmpbuf, sizeof(tmpbuf));
#  elif GETSERVBYPORT_R_ARGS == 4
      if (getservbyport_r(port, proto, &se, (void *)tmpbuf) != 0) {
        sep = NULL;
      }
#  else
      /* Lets just hope the OS uses TLS! */
      sep = getservbyport(port, proto);
#  endif
#else
      /* Lets just hope the OS uses TLS! */
#  if (defined(NETWARE) && !defined(__NOVELL_LIBC__))
      sep = getservbyport(port, (char *)proto);
#  else
      sep = getservbyport(port, proto);
#  endif
#endif
    }
    if (sep && sep->s_name) {
      /* get service name */
      name = sep->s_name;
    } else {
      /* get port as a string */
      snprintf(tmpbuf, sizeof(tmpbuf), "%u", (unsigned int)ntohs(port));
      name = tmpbuf;
    }
    name_len = ci_strlen(name);
    if (name_len < buflen) {
      /* return it if buffer big enough */
      memcpy(buf, name, name_len + 1);
    } else {
      /* avoid reusing previous one */
      buf[0] = '\0'; /* LCOV_EXCL_LINE: no real service names are too big */
    }
    return buf;
  }
  buf[0] = '\0';
  return NULL;
}

#ifdef HAVE_STRUCT_SOCKADDR_IN6_SIN6_SCOPE_ID
static void append_scopeid(const struct sockaddr_in6 *addr6, unsigned int flags,
                           char *buf, size_t buflen)
{
#  ifdef HAVE_IF_INDEXTONAME
  int is_ll;
  int is_mcll;
#  endif
  char   tmpbuf[IF_NAMESIZE + 2];
  size_t bufl;

  tmpbuf[0] = '%';

#  ifdef HAVE_IF_INDEXTONAME
  is_ll   = IN6_IS_ADDR_LINKLOCAL(&addr6->sin6_addr);
  is_mcll = IN6_IS_ADDR_MC_LINKLOCAL(&addr6->sin6_addr);
  if ((flags & CI_NI_NUMERICSCOPE) || (!is_ll && !is_mcll)) {
    snprintf(&tmpbuf[1], sizeof(tmpbuf) - 1, "%lu",
             (unsigned long)addr6->sin6_scope_id);
  } else {
    if (if_indextoname(addr6->sin6_scope_id, &tmpbuf[1]) == NULL) {
      snprintf(&tmpbuf[1], sizeof(tmpbuf) - 1, "%lu",
               (unsigned long)addr6->sin6_scope_id);
    }
  }
#  else
  snprintf(&tmpbuf[1], sizeof(tmpbuf) - 1, "%lu",
           (unsigned long)addr6->sin6_scope_id);
  (void)flags;
#  endif
  tmpbuf[IF_NAMESIZE + 1] = '\0';
  bufl                    = ci_strlen(buf);

  if (bufl + ci_strlen(tmpbuf) < buflen) {
    /* only append the scopeid string if it fits in the target buffer */
    ci_strcpy(&buf[bufl], tmpbuf, buflen - bufl);
  }
}
#endif

/* Determines if s1 ends with the string in s2 (case-insensitive) */
static char *ci_striendstr(const char *s1, const char *s2)
{
  const char *c1;
  const char *c2;
  const char *c1_begin;
  int         lo1;
  int         lo2;
  size_t      s1_len = ci_strlen(s1);
  size_t      s2_len = ci_strlen(s2);

  if (s1 == NULL || s2 == NULL) {
    return NULL;
  }

  /* If the substr is longer than the full str, it can't match */
  if (s2_len > s1_len) {
    return NULL;
  }

  /* Jump to the end of s1 minus the length of s2 */
  c1_begin = s1 + s1_len - s2_len;
  c1       = c1_begin;
  c2       = s2;
  while (c2 < s2 + s2_len) {
    lo1 = ci_tolower((unsigned char)*c1);
    lo2 = ci_tolower((unsigned char)*c2);
    if (lo1 != lo2) {
      return NULL;
    } else {
      c1++;
      c2++;
    }
  }
  /* Cast off const */
  return (char *)((size_t)c1_begin);
}

ci_bool_t ci_is_onion_domain(const char *name)
{
  if (ci_striendstr(name, ".onion")) {
    return CI_TRUE;
  }

  if (ci_striendstr(name, ".onion.")) {
    return CI_TRUE;
  }

  return CI_FALSE;
}
